plugins/proto-schema-compatibility-plugin/README.md

```diff
@@ -0,0 +1,11 @@
+# Proto Schema Compatibility Plugin
+The proto-schema-compatibility-plugin is a Maven plugin to run a backwards compatibility check on a set of protobuf schemas.
+
+The plugin utilises [protolock](https://github.com/nilslice/protolock) to generate a `proto.lock` file for each module
+that it is applied to. If a `proto.lock` file already exists for that module, then the existing file is utilised to ensure
+that any changes to the latest code are backwards compatible with the definitions in the existing lock file. As proto
+schemas evolve, it's necessary for a modules `proto.lock` changes to be committed to git.
+
+To force any breaking changes and reset the current state, it's necessary to delete the offending `proto.lock` file(s),
+however this should only be done when fields have had sufficient time to be truly obsolete, i.e. across multiple major
+version changes when rolling upgrades are no longer supported.
```

Comment:
> I see protolock has some support for eventually removing reserved fields.

Yes it does, this was my motivation for requesting [IPROTO-11](https://issues.jboss.org/browse/IPROTO-11) so that can make use of this via our annotated Pojos.



> And maybe "As proto
> schemas evolve, it's necessary for a modules `proto.lock` changes to be committed to git." is too vague, considering that non-breaking proto schema changes are additive maybe we should say that generated `proto.lock` files should be committed to git after every Final release?

I'm not sure I understand your point here. The `proto.lock` files are rewritten every time a change happens *if* the changes are backwards compatible. However, if the changes are not compatible protolock fails and the `proto.lock` file is not updated. 

It is necessary for us to commit the `proto.lock` files for every compatible change, as it will ensure that when building with the latest changes you are verifying against the current state of the schema. For final releases we don't have todo anything special, as we can get the exact contents of the `proto.lock` files by simply checking out the git tag of the release.

Reply:
My point is that we should not commit the proto.lock files on every change, because we don't need backwards compatibility with the latest snapshot or even Beta, we only need backwards compatibility with Final releases.

Reply:
So one option would be to just enable the checks for `-Pdistribution`, however I think we'll get into a situation similar to our javadocs where we only find out at release time that there is an issue and I fear that a backwards incompatible change could be much harder to fix quickly. IMO it's better to fail-fast so that the contributor is aware of the issue they're potentially introducing. How often do you think the situation will arise where an incompatible change will be introduced in one PR and fixed in another? I can't foresee a situation where the work can't be contained within a single PR.

Reply:
> So one option would be to just enable the checks for -Pdistribution

I'm all for checking compatibility every time, I just don't want to commit the changes every time.

> How often do you think the situation will arise where an incompatible change will be introduced in one PR and fixed in another?

It's entirely possible that one PR marks a field as reserved and another PR removes it. None of them would be incompatible with HEAD, but the end result would be incompatible with the previous Final version.

Reply:
Ok, you make a good point. So we could just do `protolock status` for every PR, and only call `protolock commit` when a property is set (set in `Jenkins-file`) release.